timelock-server/src/main/java/com/palantir/atlasdb/timelock/clock/ClockSkewMonitor.java

```diff
@@ -0,0 +1,176 @@
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.clock;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import javax.net.ssl.SSLSocketFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.collect.Maps;
+import com.palantir.atlasdb.http.AtlasDbHttpClients;
+import com.palantir.atlasdb.util.AtlasDbMetrics;
+import com.palantir.logsafe.SafeArg;
+
+/**
+ * ClockSkewMonitor keeps track of the system time of the other nodes in the cluster, and compares it to the local
+ * clock. It's purpose is to monitor if the other nodes' clock progress at the same pace as the local clock.
+ */
+public final class ClockSkewMonitor {
+
+    private static final long SLEEP_TIME_SECONDS = 1; // 1 s
+    private static final long NANOS_PER_SECOND = 1_000_000_000; // 10^9
+    private static final long MAX_TIME_SINCE_PREVIOUS_REQUEST_NANOS = 10 * SLEEP_TIME_SECONDS * NANOS_PER_SECOND; // 10s
+    private static final long MAX_REQUEST_TIME_NANOS = 10_000_000; // 10 ms
+    private static final long WARN_SKEW_THRESHOLD_NANOS = 10_000_000; // 10 ms
+    private static final long ERROR_SKEW_THRESHOLD_NANOS = 50_000_000; // 50 ms
+
+    private static final MetricRegistry METRIC_REGISTRY = AtlasDbMetrics.getMetricRegistry();
+
+    private final Logger log = LoggerFactory.getLogger(ClockSkewMonitor.class);
+    private final Map<String, ClockService> monitorsByServer;
+    private final Map<String, RequestTime> previousRequestsByServer;
+    private final Supplier<Boolean> shouldRunClockSkewMonitor;
+
+    public static ClockSkewMonitor create(Set<String> remoteServers, Optional<SSLSocketFactory> optionalSecurity,
+            Supplier<Boolean> shouldRunClockSkewMonitor) {
+        Map<String, ClockService> monitors = Maps.toMap(remoteServers,
+                (remoteServer) -> AtlasDbHttpClients.createProxy(optionalSecurity, remoteServer, ClockService.class));
+
+        Map<String, RequestTime> previousRequests = Maps.newHashMap();
+        for (String server : remoteServers) {
+            previousRequests.put(server, RequestTime.EMPTY);
+        }
+
+        return new ClockSkewMonitor(monitors, previousRequests, shouldRunClockSkewMonitor);
+    }
+
+    private ClockSkewMonitor(
+            Map<String, ClockService> monitorsByServer,
+            Map<String, RequestTime> previousRequestsByServer,
+            Supplier<Boolean> shouldRunClockSkewMonitor) {
+        this.monitorsByServer = monitorsByServer;
+        this.previousRequestsByServer = previousRequestsByServer;
+        this.shouldRunClockSkewMonitor = shouldRunClockSkewMonitor;
+    }
+
+    public void run() {
+        Executors.newSingleThreadScheduledExecutor().schedule(
+                () -> {
+                    try {
```

Comment:
there is a lot of indentation here - would suggest refactoring out a `runOneInterationMethod` method